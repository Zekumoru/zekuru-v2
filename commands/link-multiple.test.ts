import linkMultiple from './link-multiple';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import {
  clearSampleChannelLinks,
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';

jest.mock('./utilities/linking');
jest.mock('../cache');

describe('/link-multiple command', () => {
  afterEach(() => {
    clearSampleChannelLinks();
    jest.clearAllMocks();
  });

  it('should notify user that the command has been called outside of a server', async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await linkMultiple.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it.each([['random string'], ['<@123456789>']])(
    `%s - should notify user if they didn't provide correct channels`,
    async (stringValue) => {
      const interaction = createMockInteraction();
      interaction.options.getString.mockReturnValue(stringValue);

      await linkMultiple.execute(interaction);

      expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    }
  );

  it.each([
    ['<#000>'],
    ['<#000> <#001>'],
    ['<#000> <#001> <#002> <#003> <#004>'],
  ])(
    `should notify user that they didn't provide a translate channel or translate channels`,
    async (stringValue) => {
      const interaction = createMockInteraction();
      interaction.options.getString.mockReturnValue(stringValue);

      await linkMultiple.execute(interaction);

      expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    }
  );

  it(`should notify user that they only provided one translate channel`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('<#100>');

    await linkMultiple.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user that they reached linking limit`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue(
      '<#100> <#200> <#300> <#400>'
    );

    await linkMultiple.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toHaveLength(0); // 100
    expect(sampleChannelLinks[1].links).toHaveLength(0); // 200
    expect(sampleChannelLinks[2].links).toHaveLength(0); // 300
    expect(sampleChannelLinks[3].links).toHaveLength(0); // 400
  });

  it(`should link channels up to the limit`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('<#100> <#200> <#300>');

    await linkMultiple.execute(interaction);

    expect(sampleChannelLinks[0].links).toEqual([
      sampleTranslateChannels[1],
      sampleTranslateChannels[2],
    ]); // 100 linked to: 200, 300
    expect(sampleChannelLinks[1].links).toEqual([
      sampleTranslateChannels[0],
      sampleTranslateChannels[2],
    ]); // 200 linked to: 100, 300
    expect(sampleChannelLinks[2].links).toEqual([
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ]); // 300 linked to: 100, 200
  });

  it(`it shouldn't link anymore if channels are already linked`, async () => {
    sampleChannelLinks[0].links = [
      sampleTranslateChannels[1],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[1].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[2].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ] as any;
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('<#100> <#200> <#300>');

    await linkMultiple.execute(interaction);

    expect(sampleChannelLinks[0].links).toEqual([
      sampleTranslateChannels[1],
      sampleTranslateChannels[2],
    ]); // 100 linked to: 200, 300
    expect(sampleChannelLinks[1].links).toEqual([
      sampleTranslateChannels[0],
      sampleTranslateChannels[2],
    ]); // 200 linked to: 100, 300
    expect(sampleChannelLinks[2].links).toEqual([
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ]); // 300 linked to: 100, 200
  });

  it.each([
    ['<#100> <#400>'],
    ['<#200> <#300> <#400>'],
    ['<#100> <#200> <#300> <#400>'],
  ])(
    `%s - it shouldn't link anymore if the limit has already been reached`,
    async (stringValue) => {
      sampleChannelLinks[0].links = [
        sampleTranslateChannels[1],
        sampleTranslateChannels[2],
      ] as any;
      sampleChannelLinks[1].links = [
        sampleTranslateChannels[0],
        sampleTranslateChannels[2],
      ] as any;
      sampleChannelLinks[2].links = [
        sampleTranslateChannels[0],
        sampleTranslateChannels[1],
      ] as any;
      const interaction = createMockInteraction();
      interaction.options.getString.mockReturnValue(stringValue);

      await linkMultiple.execute(interaction);

      expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
      expect(sampleChannelLinks[0].links).toEqual([
        sampleTranslateChannels[1],
        sampleTranslateChannels[2],
      ]); // 100 linked to: 200, 300
      expect(sampleChannelLinks[1].links).toEqual([
        sampleTranslateChannels[0],
        sampleTranslateChannels[2],
      ]); // 200 linked to: 100, 300
      expect(sampleChannelLinks[2].links).toEqual([
        sampleTranslateChannels[0],
        sampleTranslateChannels[1],
      ]); // 300 linked to: 100, 200
    }
  );
});
