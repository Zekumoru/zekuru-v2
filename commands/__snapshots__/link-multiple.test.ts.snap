// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/link-multiple command <#100> <#200> <#300> <#400> - it shouldn't link anymore if the limit has already been reached 1`] = `
{
  "content": "<#100>, <#200>, <#300> and <#400> are **not linked** because you already reached the linking limit of 3!",
}
`;

exports[`/link-multiple command <#100> <#400> - it shouldn't link anymore if the limit has already been reached 1`] = `
{
  "content": "<#100> and <#400> are **not linked** because you already reached the linking limit of 3!",
}
`;

exports[`/link-multiple command <#200> <#300> <#400> - it shouldn't link anymore if the limit has already been reached 1`] = `
{
  "content": "<#200>, <#300> and <#400> are **not linked** because you already reached the linking limit of 3!",
}
`;

exports[`/link-multiple command <@123456789> - should notify user if they didn't provide correct channels 1`] = `
{
  "content": "Cannot link, you didn't provide any channels!",
}
`;

exports[`/link-multiple command random string - should notify user if they didn't provide correct channels 1`] = `
{
  "content": "Cannot link, you didn't provide any channels!",
}
`;

exports[`/link-multiple command should notify user that the command has been called outside of a server 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/link-multiple command should notify user that they didn't provide a translate channel or translate channels 1`] = `
{
  "content": "Cannot link, you didn't provide any translate channels!
<#000> is not linked because it is not set to any language. Please use the \`/set\` command to specify its language.",
}
`;

exports[`/link-multiple command should notify user that they didn't provide a translate channel or translate channels 2`] = `
{
  "content": "Cannot link, you didn't provide any translate channels!
<#000> and <#001> are not linked because they are not set to any language. Please use the \`/set\` command to specify their languages.",
}
`;

exports[`/link-multiple command should notify user that they didn't provide a translate channel or translate channels 3`] = `
{
  "content": "Cannot link, you didn't provide any translate channels!
<#000>, <#001>, <#002>, <#003> and <#004> are not linked because they are not set to any language. Please use the \`/set\` command to specify their languages.",
}
`;

exports[`/link-multiple command should notify user that they only provided one translate channel 1`] = `
{
  "content": "Cannot link, please provide more than one translate channel!",
}
`;

exports[`/link-multiple command should notify user that they reached linking limit 1`] = `
{
  "content": "<#100>, <#200>, <#300> and <#400> are **not linked** because you already reached the linking limit of 3!",
}
`;
